Cc3220sf host driver

GAnthony opened the pull request:
This PR imports the SimpleLink host driver v 1.40.00.03 (which communicates to the Network Coprocessor over dedicated SPI) plus its port to Zephyr for the CC3220SF_LaunchXL.

This is the first step to enable BSD socket offload over WiFi for the CC3220SF.

This is a replacement for the stale, now abandoned, PR #412, and is based on the (merged) SimpleLink SDK v 1.40.00.03 update PR #772.

Commit:
ext: simplelink: Add SimpleLink DPL porting layer to Zephyr

This ports the SimpleLink WiFi host driver via its OS adaptation
layer to Zephyr OS primitives.

This was validated using an out-of-tree SimpleLink shell
application including functions for:
* WLAN connect, disconnect and scan
* Socket: UDP server and client
* NetApp: pinging an external host.

Jira: ZEP-1958

Signed-off-by: Gil Pitney <gil.pitney@linaro.org>

Commit:
cc32xx: Redefine CMSIS IRQn_Type enum to unsigned int

Previously, calling NVIC_SetPriority(IRQn_Type irqn, ....) with
the NWP interrupt number of 171 caused a hard fault during a
subsequent svc #0 instruction during _Swap().

GNU compiler is generating a bit extension instruction (sxtb) which
converts a positive IRQ value argument to a negative value when
casting to the CMSIS IRQn_Type enum parameter type.

This generates a negative index, which then writes to an SCB
control register instead of NVIC register, causing a hard
fault later on.

This issue only occurs when passing interrupt numbers > 0x80
(eg: 171 (0xab) for the NWP) to the CMSIS NVIC apis.

The solution here is simply to redefine IRQn_Type to be an
unsigned 32 bit integer, while redefining the CMSIS IRQn_Type
enum definitions for interrupts less than zero.

Jira: ZEP-1958

Signed-off-by: Gil Pitney <gil.pitney@linaro.org>

Commit:
ext: simplelink: Import TI SimpleLink SDK host driver for CC3220SF

The SimpleLink SDK provides a WiFi host driver to access the
network coprocessor over a dedicated SPI.

This patch imports the host driver plus other driver files
required to communicate with the network coprocessor.

All files have been copied from a Linux SDK installation, without
modification (except for whitespace cleanup and file mode fixes).

Note that one of the imported files, wifi/porting/user.h, has to
be modified to enable the DPL (Driver Porting Layer) port to Zephyr.
This is provided in a follow-on DPL porting patch.

Origin: Texas Instruments SimpleLink CC3220 SDK URL:
http://www.ti.com/tool/download/SIMPLELINK-CC3220-SDK Purpose:
Provide WiFi host driver for TI CC32XX SoC

Jira: ZEP-1958

Signed-off-by: Gil Pitney <gil.pitney@linaro.org>

Commit:
ext: simplelink: Enable build of the SimpleLink host driver.

This patch enables and builds the SimpleLink host driver
for Zephyr introduced in a previous patch.

Jira: ZEP-1958

Signed-off-by: Gil Pitney <gil.pitney@linaro.org>

galak commented:
What's the issue with conflict on __errno?

GAnthony commented:
SimpleLink defines its own __errno, but has a feature to use the porting OS's definition.   This patch actually should be using Zephyr's __errno definition.  CONFIG_ERRNO=y, so there should be no conflict.

galak commented:
Can we than move the errno fixes out of the 'ext: simplelink: Add SimpleLink DPL porting layer to Zephyr' patch and into 'ext: simplelink: Enable build of the SimpleLink host driver.'?

GAnthony commented:
Actually, that would be a bit difficult, as in order to enable use of Zephyr's __errno, the dpl_set_errno() function must be defined, which is part of the SimpleLink DPL porting layer, which comes in a later commit.

When staging the commits, it seemed logical to first simply enable the build of the host driver (with CONFIG_ERRNO=n, which defaults to using SimpleLink's __errno); then later, in the DPL porting layer commit, to actually hook into Zephyr's __errno (as it's part of the OS porting task), then set CONFIG_ERRNO=y.

GAnthony commented:
So, to clarify, CONFIG_ERRNO=n is OK for the  'ext: simplelink: Enable build of the SimpleLink host driver.' commit, and creates no conflict.  It just defaults to using SimpleLink's __errno.  
The later commit, when DPL layer is ported to Zephyr, then enables using Zephyr's __errno.

galak commented on ext/hal/ti/simplelink/source/ti/drivers/net/wifi/porting/user.h:
```diff
@@ -664,7 +666,19 @@ typedef signed int _SlFd_t;
+ * to allow the SimpleLink driver.c to compile
+ * Also, supply the external errno setter function.
+ */
+#include <errno.h>
```
is there any ability to have this change be in its own file?  Trying to limit changes to imported files if at all possible.

GAnthony replied:
Actually, user.h is one of the files required to change for porting SimpleLink to another OS.  That's why it's (in latest SDK releases) placed in the wifi/**porting/** directory.

The goal is to avoid changing another imported file (driver.c), which needs to include the OS's BSD errno definition when SimpleLink's internal errno mechanism is not used.   Those changes are allowed to be made in the user.h header file (in the wifi/porting/ directory), which gets included in simplelink.h, which driver.c includes.

From driver.c:
#ifdef SL_INC_INTERNAL_ERRNO
#include <ti/drivers/net/wifi/bsd/errno.h>
#endif

So, one alternative would be to modify driver.c directly to include Zephyr's errno.h, but that is not allowed.  I'm not sure there is a different alternative.

I think this change is in the proper file.

galak replied:
ok, works for me.

GAnthony replied:
However, it would make sense to place all wifi/porting/user.h changes for the port in a separate *commit*, to take it out of the import commit.

galak replied:
Yeah, that would be helpful, having that commit message comment about user.h being expected to be modified for OS adaption would be good.

GAnthony replied:
OK, rebased and split errno changes into a separate commit.
Thanks @galak for the review.

galak reviewed (changes requested):
Can you rebase this on latest tree.

dbkinder reviewed (dismissed):
+1 for docs, but there are merge conflicts to address

Commit:
ext: simplelink: Enable SimpleLink to use Zephyr __errno

The SimpleLink host driver comes with its own definition of
__errno, which conflicts with Zephyr's definition, but has
a mechanism to enable use of the porting OS's __errno variable.

This patch enables SimpleLink to use Zephyr's __errno
via the DPL porting layer.

Jira: ZEP-1958

Signed-off-by: Gil Pitney <gil.pitney@linaro.org>

galak commented:
Can you rebase, also do we want to update to SDK/HAL version 1_40_01_00?

GAnthony commented:
Rebased.

I reviewed SDK 1.40.01.00 vs 1.40.00.03 and found no significant functionality differences.

Given the effort to update the SDK file subsets, and the fact this PR is blocking other work, I'd prefer this get merged now, and we update the SDK/HAL versions later if need arises.
